### PR TITLE
BUG: yield correct Series subclass in df.iterrows() (#13977)

### LINKED ISSUE #13977: df.iterrows() constructs Series instead of its proper subclass
#### Code Sample, a copy-pastable example if possible
```py
import pandas as pd
import numpy as np


class SubclassedSeries(pd.Series):
    pass


class SubclassedDataFrame(pd.DataFrame):
    _constructor_sliced = SubclassedSeries


df = SubclassedDataFrame(np.random.random((3, 2))
i, row = next(df.iterrows())

assert isinstance(row, SubclassedSeries)  # fails
```

#### Expected Output

`row` would be of proper overriden `_constructor_sliced` type (i.e. `SubclassedSeries`).

### PR DESCRIPTION
**Pull request by kernc**
- [x] closes #13977
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
pls read the instructions. we explicity call for tests.

**Comment by kernc** [comment]
Ah, sorry. I made sure existing tests passed. Now I also added a new one.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -693,8 +693,9 @@ def iterrows(self):
 
         """
         columns = self.columns
+        _Series = self._constructor_sliced
```
Comment (jreback) [inline pandas/core/frame.py]: call this ``klass``

File: pandas/tests/frame/test_subclass.py [hunk 2]
```diff
@@ -211,6 +211,12 @@ def test_subclass_align_combinations(self):
         tm.assertIsInstance(res2, tm.SubclassedDataFrame)
         tm.assert_frame_equal(res2, exp1)
 
+    def test_subsclass_iterrows(self):
+        # GH 13977
+        df = tm.SubclassedDataFrame({'a': [1]})
+        for i, row in df.iterrows():
+            self.assertIsInstance(row, tm.SubclassedSeries)
```
Comment (jreback) [inline pandas/tests/frame/test_subclass.py]: check these rows using ``assert_series_equal`` (IOW construct the expectation

**Comment by jreback** [comment]
minor changes. ping when green.

**Comment by kernc** [comment]
Revised. Hope it's ok.

I did notice some more `Series` and `DataFrame` calls where `self._constructor_sliced` or `self._constructor` might make sense (e.g. in `DataFrame.apply()` and submethods). What is the rationale for ever using `Series` instead of `self._constructor_sliced`?

File: pandas/tests/frame/test_subclass.py [hunk 3]
```diff
@@ -211,6 +211,12 @@ def test_subclass_align_combinations(self):
         tm.assertIsInstance(res2, tm.SubclassedDataFrame)
         tm.assert_frame_equal(res2, exp1)
 
+    def test_subclass_iterrows(self):
+        # GH 13977
+        df = tm.SubclassedDataFrame({'a': [1]})
+        for i, row in df.iterrows():
+            tm.assert_series_equal(row, df.loc[i])
```
Comment (sinhrks) [inline pandas/tests/frame/test_subclass.py]: can u also add ``assertIsInstance`` explicitly to check it is ``SubclassedSeries``? ``assert_series_equal`` only checks class identity, not check what it is.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -695,8 +695,9 @@ def iterrows(self):
 
         """
         columns = self.columns
+        klass = self._constructor_sliced
         for k, v in zip(self.index, self.values):
-            s = Series(v, index=columns, name=k)
+            s = klass(v, index=columns, name=k)
```
Comment (sinhrks) [inline pandas/core/frame.py]: call ``__finalize__`` after ``_constructor...``.

**Comment by sinhrks** [comment]
Thx for the update. Yes fixing relevant cases are appreciated (OK for separate PR).  The remaining normal constructor is mostly because of historical reason, and can be replaced to ``constructor...``.

File: pandas/core/frame.py (hunk 4 above)
Comment (kernc) [inline pandas/core/frame.py]: I took existing code as example and `__finalize__` is never called for `_constuctor_sliced`. :worried:

File: pandas/tests/frame/test_subclass.py (hunk 3 above)
Comment (kernc) [inline pandas/tests/frame/test_subclass.py]: Even though this is already tested in `test_indexing_sliced` which tests `df.loc`?

**Comment by kernc** [comment]
But what are relevant cases? If `DataFrame._constructor` must always be a subtype of `DataFrame` (and so on for `_constructor_sliced` and `_constructor_expanddim`), why not just do a mass renaming of `DataFrame` into `self._constructor` (etc.) in all methods?

Using Python's `inspect` module, one could even write a simple test ensuring no references to Series, DataFrame and Panel are hardcoded. :bulb:

**Comment by sinhrks** [comment]
It's simply because no one can take a time to do it (add tests). Note that sigatures must be changed sometimes, see #13208.

**Comment by kernc** [comment]
Thanks, will look into it.

**Comment by kernc** [comment]
Adding just `__finalize__()` made [tests no longer pass](https://travis-ci.org/pydata/pandas/jobs/152388376#L1018).

I added a RFC commit with one proposed solution. It basically makes only DataFrame inherit DataFrame's _metadata properties, Series Series', and so on. I don't know whether this opposes any established behavior / documentation. Doesn't seem to. The tests will pass. :pray:

**Comment by jreback** [comment]
the inheriting of metadata should be on a like object. However this would belong in another PR.

**Comment by sinhrks** [comment]
@kernc thx for theck. I've thought ``finalize`` has been generally called, sorry.

**Comment by kernc** [comment]
So this appears to be it for this PR. I'll be opening the other one shortly.

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by kernc** [comment]
So this is green. I'm not sure why Xcode build is marked as failed.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**